### ISSUE TITLE
Implement accessibility checker for CKEditor 5

### ISSUE THREAD

**Issue by Reinmar** [ckeditor/ckeditor5#1941]
As discussed in https://github.com/ckeditor/ckeditor5/issues/1908 we designed features in CKEditor 5 in a way which helps the users to create accessible content. However, in order to ensure that the content produced by CKEditor 5 is accessible, we'd need a similar solution to [CKEditor 4's accessibility checker](https://ckeditor.com/ckeditor-4/accessibility-checker/).

---

If you'd like to see accessibility checker for CKEditor 5 add 👍 to this post.

**Comment by mgifford** [comment]
I wouldn't recommend this @JayNathSF as it is still based on QuailJS last I checked.  Would be much better to move to [axe](https://github.com/dequelabs/axe-core), [HTML Code Sniffer](https://squizlabs.github.io/HTML_CodeSniffer/), [Tota11y](https://khan.github.io/tota11y/) or [Sa11y](https://ryersondmp.github.io/sa11y/) 

Let's not continue to ship unsupported JavaScript.

**Comment by mgifford** [comment]
Also worth checking out the [We4Authors Cluster](https://accessibilitycluster.com/) &.Test 7 - https://www.funka.com/en/accessibilitytest

**Comment by mgifford** [comment]
Also, flagging the conversation from CKEditor 4 - https://github.com/cksource/ckeditor4-plugin-a11ychecker/issues/242#issuecomment-763055511

**Comment by dylanb** [comment]
axe-core is used in CMS and editor integrations that are not open source and/or free and has features that are well designed to be used by such an integration

1. axe-core automatically turns rules on and off based on whether you are analyzing the entire HTML page or just a section of it so it should give your content editors exactly the right results depending on what you analyze.
2. axe-core allows individual rules to be run with a simple configuration and also allows individual rules to be turned off with a simple configuration that can be passed to the run function. So turning off color contrast for example is really easy if that is not something an editor can influence.

**Comment by itmaybejj** [comment]
Editoria11y maintainer here. Would love to help. 

Ideally in my opinion there would be live ambient hinting rather than a manual check button. We really could integrate certain checks and hints into the tools themselves. Most authors in my observation don't click manual check buttons; that's why I created an automatic checker. Certainly not authors who haven't been to accessibility trainings...

E.g. -- If the editor makes a whole short paragraph bold, throw a tooltip in that moment asking if they want to make it a heading. And if they are creating that heading after an h2, maybe float h2 and h3 to the top of the list, as the others are probably mistakes -- and break the list up with a "not recommended" divider! And if they create a table, create it with header rows **_by default_**; put the burden on them to remove the headers. And if they are adding an image, don't have them leave the alt field blank to mark it as decorative -- make the field required but have a [separate toggle](https://www.drupal.org/project/decorative_image_widget)  to "mark the image as decorative so it is ignored by screen readers."

Which is to say: a lot of what I am checking for is situations where the defaults and options lead people astray. I'd much rather help integrate those hints at the content creation moment than reimplement a manual checker.

**Comment by JayNathSF** [comment]
@itmaybejj great analogy of spell checking and doing accessibility check in real-time to improve utilization. 

Curious if Editoria11y is already integrated into CKEditor (my understanding is Drupal ships with it but I haven't used Drupal in a few years). 

And really exciting to see your expertise in this area and your offer to help! 

Let us know if we can help in any way.

**Comment by mgifford** [comment]
As far as a clarification @JayNathSF Drupal doesn't ship with it, but there is a module to add it in to CKEditor for Drupal. 

My goal would be to have something like this integrated into CKEditor 5 and have that enabled for Drupal Core at some point.

**Comment by itmaybejj** [comment]
Yeah; Editoria11y's Drupal integration triggers **_after_** the page renders (outside of the checker). I consider that a [stopgap step forward](https://accessibility.princeton.edu/news/introducing-editoria11y).

That said...my education in assembling Editoria11y was that checkers all follow a similar pattern inside the black box: 

1. Traverse the DOM looking for specific elements
2. Run some `if` statements (most are no more complicated than "is that there alt attribute blank or null?")
3. Inject markup into the page. 

Which is to say: once you get past the "scary black box" factor, there's no need to become _dependent_ on an externally library. You will have to write your own triggers and popup integrations anyway (1 & 2), so it might make sense to just think about what **_tests_** are relevant to content inside a CKEditor box (it's less than a quarter of AXE, trust me), and then yoink and adapt those specific `if` statements from an open source library to directly integrate into your plugin. Then you have no dependencies and can tweak it to fit your editing-in-progress markup.

Performance is such an issue in these things btw -- I spent a **ton** of time optimizing to not have my library make the page janky -- but my tests only run once, not over and over while a user is editing. In your case -- there's no need to run **_all_** the tests over and over again. I imagine it would make sense to run one full sweep when the editor opens, and then monitor the cursor position and only re-run relevant tests for the just-active element type when the user clicks, taps or presses enter, spacebar or tab; maybe combined with a bit of a debounce so it doesn't run while they are typing fast.

And then you could use or parallel CKEditor's native tooltip architecture to do something on the fly like...
![ckeditor-alert-mockup](https://user-images.githubusercontent.com/4519454/112156695-de9ee580-8bbc-11eb-95bf-1ca614e397c3.png)

**Comment by olets** [comment]
> And if they are creating that heading after an h2, maybe float h2 and h3 to the top of the list, as the others are probably mistakes -- and break the list up with a "not recommended" divider! 

I recognize this probably isn't the place to write out a wish list, but since this specific thing was mentioned:

This would be a dream come true for my team. It's something we talk about and work on ad hoc solutions for repeatedly. Support for it would be strong point in favor of using CKEditor.

**Comment by itmaybejj** [comment]
It's CK4 -- dunno how much code translates -- but this group already has working code for it: https://github.com/a11yfirst/ckeditor-plugins/tree/master/plugins/a11yfirst-headings

**Comment by mgifford** [comment]
Also worth noting this approach from the We4Authors Cluster https://accessibilitycluster.com/main-results/live-testing-while-authoring

**Comment by mattsimpson** [comment]
This is still a really important feature for us, and something we really liked and benefited from in CKEditor 4. Is there any plans on reintroducing this functionality?

**Comment by itmaybejj** [comment]
I have finished the rewrite of Editoria11y and documented how to install it as a [CKeditor5 Accessibility Checker](https://editoria11y.princeton.edu/2024/11/ckeditor5-accessibility-checker/).

**Comment by 108signals** [comment]
@itmaybejj amazing! Thank you so much for all of the work in developing this. Can't wait to try this out. Lmk if you want to chat about our use case (we serve public sector agencies)

**Comment by itmaybejj** [comment]
I'll definitely want user feedback. Trying to slot tips in between editor controls and underline-based grammar checker plugins without breaking either turned out to be quite challenging.
